### ISSUE TITLE
Life cycle of the OpenSSL 3.0 FIPS module

### ISSUE THREAD

**Issue by mspncp** [openssl/openssl#13693]
Currently, the OpenSSL FIPS module is installed unconditionally by the `install_sw` target together with the other providers. The `install_fips` target only creates and installs the configuration file.  If the installation happens to be a service update installation, then the `openssl install` command will invalidate a previously installed FIPS module.

IIRC, it was the plan to support service updates (bug fixes) of the 'regular' OpenSSL software independently from the FIPS provider. The current installation procedure does not seem to be fit to support that goal. Are we going to fix this and provide a concept how to deal the life cycle and versioning of providers, or are we going to leave the nasty details to the distributors?

One possible step in the right direction would be to exclude the FIPS provider from the `install_sw` target and use the `install_fips` target to for both the provider and the configuration file.

_Originally posted in https://github.com/openssl/openssl/pull/13684#issuecomment-745609203_

**Comment by mspncp** [comment]
_(@levitte's answer, copied from https://github.com/openssl/openssl/pull/13684#issuecomment-745746264)_

> IIRC, it was the plan to support service updates (bug fixes) of the 'regular' OpenSSL software independently from the FIPS provider. The current installation procedure does not seem to be fit to support that goal. Are we going to fix this and provide a concept how to deal the life cycle and versioning of providers, or are we going to leave the nasty details to the distributors?
> 
> One possible step in the right direction would be to exclude the FIPS provider from the `install_sw` target and use the `install_fips` target to for both the provider and the configuration file.

Good questions.  It does make sense to install it separately, but [sigh] another special case.  Oh well.

**Comment by mspncp** [comment]
> Good questions. It does make sense to install it separately, but [sigh] another special case. Oh well.

Instead of hacking a special case for fips, couldn't you just add the information about the appropriate install target in a generic way as new entry to the build info file?

**~/src/openssl/providers/fips/build.info**
```
SOURCE[../fips]=fipsprov.c self_test.c self_test_kats.c
INCLUDE[../fips]=../implementations/include ../common/include ../..
```

**Comment by levitte** [comment]
> Instead of hacking a special case for fips, couldn't you just add the information about the appropriate install target in a generic way as new entry to the build info file (for all providers)?

We already have that, in form of a build.info attribute ("fips"), see this line:

https://github.com/openssl/openssl/blob/74cd923a78b490d46af9c3dc5c8dc7a741c5e576/providers/build.info#L143

It still needs to be treated as a *separate* module, away from all other possible modules.

**Comment by levitte** [comment]
I did a quick hack to at least put the idea in display, currently Unix only.  See #13969

**Comment by mspncp** [comment]
Looks reasonable at first sight. To take a deeper look, I'll probably simply cherry-pick your commit to #13684 to test it together with my changes.

**Comment by t8m** [comment]
Are we going to move this forward?

**Comment by mspncp** [comment]
> Are we going to move this forward?

Hopefully #13684 separates the 'normal' installation and the installation of the FIPS provider well enough that this issue can be considered resolved as soon as #13684 gets merged. However, I would appreciate some extra reasoning and testing by  by @openssl/committers to make sure that delayed updates of the FIPS provider (after an update of the core libary and the default provider) will work as expected in all corner cases.
